bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Display.java

```diff
@@ -5215,79 +5215,65 @@ static String withCrLf (String string) {
 	int length = string.length ();
 	if (length == 0) return string;
 
-	/*
-	* Check for an LF or CR/LF and assume the rest of
-	* the string is formated that way.  This will not
-	* work if the string contains mixed delimiters.
-	*/
+	/* If there is no LF - return string */
 	int i = string.indexOf ('\n', 0);
 	if (i == -1) return string;
-	if (i > 0 && string.charAt (i - 1) == '\r') {
-		return string;
-	}
 
-	/*
-	* The string is formatted with LF.  Compute the
-	* number of lines and the size of the buffer
-	* needed to hold the result
-	*/
-	i++;
-	int count = 1;
-	while (i < length) {
-		if ((i = string.indexOf ('\n', i)) == -1) break;
-		count++; i++;
-	}
-	count += length;
+	/*Create String Builder for return string*/
+	StringBuilder result = new StringBuilder();
+	int j = 0;
+
+	/*If there is just LF, without CR before it,
+	add CR before LF*/
 
-	/* Create a new string with the CR/LF line terminator. */
-	i = 0;
-	StringBuilder result = new StringBuilder (count);
 	while (i < length) {
-		int j = string.indexOf ('\n', i);
-		if (j == -1) j = length;
-		result.append (string.substring (i, j));
-		if ((i = j) < length) {
-			result.append ("\r\n"); //$NON-NLS-1$
-			i++;
+		result.append(string.substring(j, i));
+		if(string.charAt(i - 1) == '\r'){
+			result.append('\n');
+		}
+		else {
+			result.append("\r\n");//$NON-NLS-1$
+		}
+		j = i + 1;
+		i = string.indexOf('\n', j);
+		if (i == -1) {
+			i = length;
+			result.append(string.substring(j));
 		}
 	}
-	return result.toString ();
+	return result.toString();
```

Comment:
I did not took a detailed look in the original coding, but according to ChatGTP, we could just use a regex to normalize the String. See the coding:
```
/**
     * Normalize line breaks in a string to Windows line breaks (\r\n)
     *
     * @param input The input string
     * @return The string with normalized line breaks
     */
    public static String normalizeLineBreaks(String input) {
        // Replace \r\n, \r, or \n with \r\n
        return input.replaceAll("(\r\n|\r|\n)", "\r\n");
    }
```

This seems pretty easy and clean. Maybe you can give it a try.